test/TestUserAcceptance.hs

```diff
@@ -295,59 +242,69 @@ cleanUpTmuxSession sessionname = do
 
 
 -- | Run all application steps in a session defined by session name.
-tmuxSession :: String -> [ApplicationStep] -> TestTree
-tmuxSession tcname xs =
-  withResource setUp tearDown $
-  \_ -> testCaseSteps tcname $ \step -> runSteps step xs
-
-runSteps :: (String -> IO ()) -> [ApplicationStep] -> IO ()
-runSteps stepfx steps =
-    mapM_
-        (\a ->
-              do stepfx (asDescription a)
-                 out <- performStep "purebredtest" a
-                 d <- lookupEnv "DEBUG"
-                 when (isJust d) $ hPutStr stderr ("\n\n" ++ asDescription a ++ "\n\n" ++ out)
-                 ((asAssertInOutput a) out (asExpected a)))
-        steps
-
-performStep :: String -> ApplicationStep -> IO (String)
-performStep sessionname (ApplicationStep keys _ asLiteral expect _) = do
-    callProcess "tmux" $ communicateSessionArgs keys asLiteral
-    baton <- newEmptyMVar
-    out <- waitForString baton sessionname expect
-    _ <- takeMVar baton
-    pure out
+withTmuxSession :: TestName -> ((String -> IO ()) -> ReaderT Env IO ()) -> TestTree
+withTmuxSession tcname testfx =
+    withResource setUp tearDown $
+      \env -> testCaseSteps tcname $ \stepfx -> env >>= runReaderT (testfx stepfx)
+
+sendKeys :: String -> String -> ReaderT Env IO (String)
+sendKeys keys expect = do
+    sessionname <- getSessionName
+    liftIO $ do callProcess "tmux" $ communicateSessionArgs keys False
+                wait sessionname expect
+
+sendLiteralKeys :: String -> ReaderT Env IO (String)
+sendLiteralKeys keys = do
+    sessionname <- getSessionName
+    liftIO $ do callProcess "tmux" $ communicateSessionArgs keys True
+                wait sessionname keys
+
+capture :: ReaderT Env IO (String)
+capture = do
+  sessionname <- getSessionName
+  liftIO $ readProcess "tmux" ["capture-pane", "-e", "-p", "-t", sessionname] []
+
+getSessionName :: ReaderT Env IO (String)
+getSessionName = view (_3 . ask)
 
 holdOffTime :: Int
 holdOffTime = 10^6
 
+debugOutput :: String -> IO ()
+debugOutput out = do
+  d <- lookupEnv "DEBUG"
+  when (isJust d) $ hPutStr stderr ("\n\n" ++ out)
+
+wait :: String -> String -> IO (String)
+wait sessionname substr = do
+  baton <- newEmptyMVar
+  out <- waitForString baton sessionname substr 0
+  i <- takeMVar baton
+  assertBool ("expected: '" ++ substr ++ "' but got:\n\n " ++ out) (i == 1)
+  pure out
+
 -- | wait for the application to render a new interface which we determine with
 --   a given substring. If the expected substring is not in the captured pane,
 --   wait a bit and try again.
-waitForString :: MVar String -> String -> String -> IO (String)
-waitForString baton sessionname substr = do
+waitForString :: MVar Int -> String -> String -> Int -> IO (String)
+waitForString baton sessionname substr i = do
```

Comment:
I think it would be better to make the Int arg a "countdown", and put the number of desired attempts at the call site.
So something like:
```haskell
waitForString :: Int {- attempts -} -> MVar Int -> String -> String -> IO (String)
waitForString n baton sessionname substr =
  readProcess "tmux" ["capture-pane", "-e", "-p", "-t", sessionname] [] >>= checkPane
  where
    checkPane out
      | substr `isInfixOf` out = putMVar baton 1 $> out
      | n <= 0 = putMVar baton 0 $> out                     -- no more attempts
      | otherwise = threadDelay holdOffTime *> waitForString baton sessionname substr (n - 1)
```

Personally I think this is easier to follow with the guards.

Then you would call it like:
```haskell
  out <- waitForString baton sessionname substr 5 {- max tries -}
```

If you ever do need to get back the "screenshot for each attempt" it is a straightforward refactor;
you just accumulate the screenshots in a list or something.

Reply:
AAahh guards... of course. Why haven't I thought of them. Cheers! I like the countdown.